apps/bot_manager/lib/bot_state_machine.ex

```diff
@@ -3,63 +3,233 @@ defmodule BotManager.BotStateMachine do
+    case next_state do
+      :moving ->
+        direction = maybe_switch_direction(bot_player, bot_state_machine)
 
-      closest_player.distance < 50 ->
-        {:move, create_random_direction()}
+        %{
+          action: determine_player_move_action(bot_player, bot_state_machine, direction),
+          bot_state_machine: bot_state_machine
+        }
 
-      closest_player.distance <= random_distance ->
-        {:attack, closest_player.direction}
+      :aggresive ->
+        use_skill(%{
+          bot_player: bot_player,
+          bot_state_machine: bot_state_machine,
+          game_state: game_state,
+          attack_blocked: attack_blocked
+        })
 
-      true ->
-        {:move, create_random_direction()}
+      :running_away ->
+        ran_away(bot_player, game_state, bot_state_machine)
     end
   end
 
-  def decide_action(_), do: :stand
+  def decide_action(%{bot_state_machine: bot_state_machine}),
+    do: %{action: :stand, bot_state_machine: bot_state_machine}
+
+  def use_skill(%{attack_blocked: true, bot_player: bot_player, bot_state_machine: bot_state_machine}),
+    do: %{action: {:move, bot_player.direction}, bot_state_machine: bot_state_machine}
+
+  def use_skill(%{
+        game_state: game_state,
+        bot_player: bot_player,
+        bot_state_machine: bot_state_machine
+      }) do
+    players_with_distances = map_directions_to_players(game_state, bot_player, @vision_range)
+
+    if Enum.empty?(players_with_distances) do
+      direction = maybe_switch_direction(bot_player, bot_state_machine)
+
+      %{
+        action: determine_player_move_action(bot_player, bot_state_machine, direction),
+        bot_state_machine: bot_state_machine
+      }
+    else
+      cond do
+        bot_state_machine.progress_for_ultimate_skill >= bot_state_machine.cap_for_ultimate_skill ->
+          bot_state_machine =
+            Map.put(
+              bot_state_machine,
+              :progress_for_ultimate_skill,
+              bot_state_machine.progress_for_ultimate_skill - bot_state_machine.cap_for_ultimate_skill
+            )
+            |> Map.put(:state, :attacking)
+
+          direction =
+            if Enum.empty?(players_with_distances) do
+              bot_player.direction
+            else
+              closest_player = Enum.min_by(players_with_distances, & &1.distance)
+              closest_player.direction
+            end
+
+          %{action: {:use_skill, @skill_2_key, direction}, bot_state_machine: bot_state_machine}
+
+        bot_state_machine.progress_for_basic_skill >= bot_state_machine.cap_for_basic_skill ->
+          bot_state_machine =
+            Map.put(
+              bot_state_machine,
+              :progress_for_basic_skill,
+              bot_state_machine.progress_for_basic_skill - bot_state_machine.cap_for_basic_skill
+            )
+            |> Map.put(:progress_for_ultimate_skill, bot_state_machine.progress_for_ultimate_skill + 1)
```

Comment:
What's this progress about?

Reply:
Oh this is progress is used to make the bots somehow accumulate energy to use a skill, it is just a gimmick. Same for the ultimate